### PR TITLE
SDIT-1234 Add integration tests for upcoming prisoner exclusion work

### PR DESCRIPTION
**Pull request by mikehalmamoj**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/OffenderProgramProfile.kt [hunk 1]
```diff
@@ -57,7 +57,7 @@ data class OffenderProgramProfile(
 
   @ManyToOne(fetch = FetchType.LAZY)
   @JoinColumn(name = "CRS_ACTY_ID")
-  val courseActivity: CourseActivity? = null,
+  val courseActivity: CourseActivity,
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/jpa/OffenderProgramProfile.kt:60]: One of those nullable columns in NOMIS that is never null - saves me `!!`ing all unnecessarily.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/activities/AllocationResourceIntTest.kt [hunk 2]
```diff
@@ -80,7 +85,13 @@ class AllocationResourceIntTest : IntegrationTestBase() {
...
 
   private fun String.withProgramStatusCode(programStatusCode: String?) =
-    replace(""""programStatusCode": "ALLOC"""", programStatusCode?.let { """"programStatusCode": "$programStatusCode"""" } ?: """"ignored": "ignored"""") // hack so we don't have to worry about trailing comma on previous line
+    replace(""""programStatusCode": "ALLOC",""", programStatusCode?.let { """"programStatusCode": "$programStatusCode",""" } ?: "")
+
+  private fun String.withPrisonerExclusions(exclusions: List<Pair<String, String?>>) =
+    replace(
+      """"prisonerExclusions": []""",
+      """"prisonerExclusions": [${exclusions.joinToString(", ") { ex -> """{"dayOfWeek": "${ex.first}", "slotCategory": ${ex.second?.let { "\"$it\"" } ?: ""}}""" }}]""",
+    )
```
Comment (mikehalmamoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/activities/AllocationResourceIntTest.kt:94]: I've not implemented the request yet but I think it's going to look like:
```
...
"prisonerExclusions": [
  { 
    "dayOfWeek": "MON",
    "slotCategory": "AM"
  },
  { 
    "dayOfWeek": "TUE",
    "slotCategory": null
  }
]
```

File: src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/activities/AllocationResourceIntTest.kt [hunk 3]
```diff
@@ -313,6 +324,28 @@ class AllocationResourceIntTest : IntegrationTestBase() {
           assertThat(it).contains("INVALID")
         }
     }
+
+    @Test
+    @Disabled("WIP")
```
Comment (mikehalmamoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/activities/AllocationResourceIntTest.kt:329]: Turned all the tests off for now, just wanted to get them down first to keep future PRs smaller

File: src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/helper/builders/OffenderActivityExclusion.kt [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+import org.springframework.stereotype.Component
+import uk.gov.justice.digital.hmpps.nomisprisonerapi.jpa.CourseActivity
+import uk.gov.justice.digital.hmpps.nomisprisonerapi.jpa.OffenderActivityExclusion
+import uk.gov.justice.digital.hmpps.nomisprisonerapi.jpa.OffenderBooking
+import uk.gov.justice.digital.hmpps.nomisprisonerapi.jpa.OffenderProgramProfile
+import uk.gov.justice.digital.hmpps.nomisprisonerapi.jpa.SlotCategory
+import uk.gov.justice.digital.hmpps.nomisprisonerapi.jpa.WeekDay
+
+@DslMarker
+annotation class OffenderActivityExclusionDslMarker
```
Comment (mikehalmamoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/helper/builders/OffenderActivityExclusion.kt:12]: The new exclusion DSL, which is a child of `CourseAllocation`

**Review by petergphillips — APPROVED** [review]
LGTM

**Pull request merged**